resilience4j-ratpack/src/main/java/io/github/resilience4j/ratpack/circuitbreaker/CircuitBreakerTransformer.java

```diff
@@ -77,6 +77,7 @@ public void error(Throwable throwable) {
 
                     @Override
                     public void complete() {
+                        circuitBreaker.releasePermission();
```

Comment:
`releasePermission` should only be called, if downstream obtained a permission to upstream, but upstream was cancelled afterwards.
If the upstream completes successfully, we should call circuitbreaker.onSuccess instead.
If upstream handles the error and therefore onComplete is invoked, we should assume that the call has completed successfully. We can't assume that the call was cancelled.

Have a look at the RxJava2 `CompletableCircuitBreaker` or Reactor `CircuitBreakerSubscriber` implementatio

Reply:
Ratpack's name `complete` is very misleading here. Effectively it's a cancellation of the Promise: https://ratpack.io/manual/current/api/ratpack/exec/Downstream.html#complete--

Basically if the upstream execution had an error and swallowed it, then it calls `complete` on all the executions downstream of it to notify them that there's no work to do since it will not provide any data to them.

Reply:
Ok, then its fine